src/dialog/provider.js

```diff
@@ -18,107 +18,73 @@ export default class FlowProvider extends EventEmitter2 {
-      }
+      const uiEq = JSON.parse(await this.ghostManager.readFile(this.flowsDir, this._uiPath(flowPath)))
 
       Object.assign(flow, { links: uiEq.links })
 
       // Take position from UI files or create default position
-      const unplacedNodes = []
-      flow.nodes.forEach(node => {
-        const uiNode = _.find(uiEq.nodes, { id: node.id }) || {}
-
-        Object.assign(node, uiNode.position)
-
-        if (_.isNil(node.x) || _.isNil(node.y)) {
-          unplacedNodes.push(node)
+      let unplacedIndex = -1
+      flow.nodes = flow.nodes.map(node => {
+        const position = _.get(_.find(uiEq.nodes, { id: node.id }), 'position')
+        const placingStep = 250
+        unplacedIndex = position ? unplacedIndex : unplacedIndex + 1
+
+        return {
+          ...node,
+          x: position ? position.x : 50 + unplacedIndex * placingStep,
+          y: position ? position.y : (_.maxBy(flow.nodes, 'y') || { y: 0 }).y + placingStep
         }
       })
 
-      const unplacedY = (_.maxBy(flow.nodes, 'y') || { y: 0 }).y + 250
-      let unplacedX = 50
-
-      unplacedNodes.forEach(node => {
-        node.y = unplacedY
-        node.x = unplacedX
-        unplacedX += 250
-      })
-
-      return flows.push({
-        location: file,
-        version: flow.version,
-        name: flow.name,
+      return {
+        name: flowPath,
+        location: flowPath,
         nodes: _.filter(flow.nodes, node => !!node),
-        catchAll: flow.catchAll,
-        startNode: flow.startNode,
-        links: flow.links,
-        skillData: flow.skillData
-      })
+        ..._.pick(flow, 'version', 'catchAll', 'startNode', 'links', 'skillData')
+      }
     })
 
-    return flows
+    return flows.filter(Boolean)
   }
 
   async saveFlows(flows) {
     const flowsToSave = await Promise.mapSeries(flows, flow => this._prepareSaveFlow(flow))
 
     for (const { flowPath, uiPath, flowContent, uiContent } of flowsToSave) {
-      if (flowPath.includes('/')) {
-        mkdirp.sync(path.dirname(flowPath))
-      }
-
-      fs.writeFileSync(flowPath, JSON.stringify(flowContent, null, 2))
-      fs.writeFileSync(uiPath, JSON.stringify(uiContent, null, 2))
+      this.ghostManager.upsertFile(this.flowsDir, flowPath, JSON.stringify(flowContent, null, 2))
+      this.ghostManager.upsertFile(this.flowsDir, uiPath, JSON.stringify(uiContent, null, 2))
     }
 
-    const flowsDir = path.resolve(this.projectLocation, this.botfile.flowsDir || './flows')
-
-    const searchOptions = { cwd: flowsDir }
-    const flowFiles = await Promise.fromCallback(callback => glob('**/*.flow.json', searchOptions, callback))
-
-    flowFiles
-      .map(fileName => path.resolve(flowsDir, './' + fileName))
-      .filter(filePath => !flowsToSave.find(flow => flow.flowPath === filePath || flow.uiPath === filePath))
-      .map(filePath => fs.unlinkSync(filePath))
+    const flowFiles = await this.ghostManager.directoryListing(this.flowsDir, '.json')
```

Comment:
This can be optimized, currently you're reading all files from the DB and then filtering. If we make the listing method slightly more flexible with the filter we can avoid it. Though it's not critical as we will rarely have more than 100 records for the given folder

Reply:
OK, optimized that. )